### PR TITLE
ARROW-13509: [C++] Take kernel with empty inputs

### PR DESCRIPTION
**Pull request by aucahuasi**
This PR makes the Take kernel pass the type from the input arrays when constructing the chunked array.
https://issues.apache.org/jira/browse/ARROW-13509

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 1]
```diff
@@ -2035,7 +2035,7 @@ Result<std::shared_ptr<ChunkedArray>> TakeCC(const ChunkedArray& values,
     ARROW_ASSIGN_OR_RAISE(new_chunks[i],
                           Concatenate(current_chunk->chunks(), ctx->memory_pool()));
   }
-  return std::make_shared<ChunkedArray>(std::move(new_chunks));
+  return std::make_shared<ChunkedArray>(std::move(new_chunks), values.type());
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:2038]: Thanks for fixing this!

There's a similar case down on line 2051 here - that should probably be fixed too.

We should also add test cases to make sure this doesn't regress.

Comment (aucahuasi) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:2038]: Thanks for the review @lidavidm I'll address your comments cc @westonpace

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 2]
```diff
@@ -996,6 +996,16 @@ TEST_F(TestTakeKernel, InvalidIndexType) {
                                          "[0.0, 1.0, 0.1]", &arr));
 }
 
+TEST_F(TestTakeKernel, TakeEmptyIndices) {
+  Datum out;
+  auto dat = Datum(std::make_shared<ChunkedArray>(ArrayFromJSON(int8(), "[]")));
+  auto idx = Datum(std::make_shared<ChunkedArray>(ArrayVector(), int32()));
+  auto ret = Take(dat, idx).Value(&out);
+  ASSERT_OK(ret);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1004]:
```suggestion
TEST_F(TestTakeKernel, TakeEmptyIndices) {
  Datum dat = ChunkedArrayFromJSON(int8(), {"[]"});
  Datum idx = ChunkedArrayFromJSON(int32(), {});
  ASSERT_OK_AND_ASSIGN(auto out, Take(dat, idx));
```

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/vector_selection_test.cc:1004]: This can be simplified a bit.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/vector_selection_test.cc:1004]: I think we should test both branches above too. This tests the changes in TestCC, we can add a test for TestAC that calls Take with an Array and an empty (chunk-less) ChunkedArray.

**Review by lidavidm — APPROVED** [review]
LGTM. I left a suggestion for the tests since we have some helper functions to make those cases easier to write.

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc (hunk 2 above)
Comment (aucahuasi) [reply cpp/src/arrow/compute/kernels/vector_selection_test.cc]: Thanks David!

**Comment by aucahuasi** [comment]
> LGTM. I left a suggestion for the tests since we have some helper functions to make those cases easier to write.

I applied the suggestions, but let me know if I forgot something thanks!

**Comment by lidavidm** [comment]
The AppVeyor failure looks like a flake (s3fs on windows being flaky). I kicked off the 'Dev PR' pipeline again though I expect that to also be a fluke.

**Pull request closed without merging**